### PR TITLE
PLT-8077 - utxo endpoint

### PR DESCRIPTION
**Pull request by willjgould**
* Added machinery which allows us to use `MonadError` in `SQLiteAggregateQuery` `Queryable` instances (even though I'm not using it for this)
* Added a `getUtxosFromAddress` endpoint
* Fixed a typo in `--help` output
* Added golden tests for the new endpoint

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense and have useful messages
    - [ ] Important changes are reflected in changelog.d of the affected packages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Targeting main unless this is a cherry-pick backport
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [ ] If relevant, reference the ADR in the PR and reference the PR in the ADR
    - [x] Reviewer requested

### CODE REVIEW

File: legacy/marconi-chain-index-legacy/src/Marconi/ChainIndex/Legacy/Indexers/Utxo.hs [hunk 1]
```diff
@@ -255,13 +255,19 @@ instance Ord Utxo where
 instance FromJSON Utxo where
```
Comment (willjgould) [inline legacy/marconi-chain-index-legacy/src/Marconi/ChainIndex/Legacy/Indexers/Utxo.hs:255]: Accidental auto-formatting change, happy to change if you want

File: marconi-chain-index/src/Marconi/ChainIndex/Api/JsonRpc/Endpoint/Utxo/Types.hs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        (lo, hi) <- buildInterval v
+        GetUtxosFromAddressParams
+          <$> v
+            .: "address"
+          <*> pure lo
+          <*> pure hi
+     in
+      Aeson.withObject "GetUtxosFromAddressParams" parse
+    where
+      buildInterval v = do
```
Comment (willjgould) [inline marconi-chain-index/src/Marconi/ChainIndex/Api/JsonRpc/Endpoint/Utxo/Types.hs:44]: I thought about this some more. I think this is more correct than a QueryError in the Queryable instance. The reason is, these constraints are detailed by the specification. As a result, I think this genuinely is a parsing error. Happy to talk more about this if you disagree.

File: marconi-chain-index/src/Marconi/ChainIndex/Api/JsonRpc/Endpoint/Utxo.hs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+  res <- join <$> traverse getUtxoQueryInputHandler (mapGetBurnTokenEventsQuery addresses query)
+  pure $ mapGetBurnTokenEventsResult =<< res
+  where
+    mapGetBurnTokenEventsQuery
+      :: [C.AddressAny] -> GetUtxosFromAddressParams -> Either (JsonRpcErr String) UtxoQueryInput
+    mapGetBurnTokenEventsQuery
+      _ -- targetAddresses
+      (GetUtxosFromAddressParams addr createdAtAfter unspentBefore) = do
+        mappedAddr <- addressMapping (pack addr)
+        -- TODO I can't actually see where we used to do this, and I get this error when running `run-pre-prod-queries`. Am I missing something?
```
Comment (willjgould) [inline marconi-chain-index/src/Marconi/ChainIndex/Api/JsonRpc/Endpoint/Utxo.hs:107]: Note that the reason for this TODO is the spec references an error `The 'address' param value must belong to the provided target addresses.`, but I can't see where we were surfacing this previously. I implemented the logic that I presumed would be the source of the error, but it caused the error to always be thrown with my current setup.

Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Api/JsonRpc/Endpoint/Utxo.hs:107]: The filtering list can be passed when we start `marconi-chain-index`, it's the one exposed by the `TargetAddresses` endpoint. If no list is given, we index all the addresses and can query them all.

File: marconi-chain-index/src/Marconi/ChainIndex/Api/JsonRpc/Endpoint/Utxo/Types.hs (hunk 2 above)
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Api/JsonRpc/Endpoint/Utxo/Types.hs:44]: The constraints are detailed by the specification but they aren't enforced by the types themselves (in the specification).
Although, as we can enforce it on the Haskell side, it's probably the way to go.

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/UtxoQuery.hs [hunk 4]
```diff
@@ -282,9 +290,14 @@ baseQuery =
   LEFT JOIN blockInfo.blockInfo spentBlockInfo ON spentBlockInfo.slotNo = futureSpent.slotNo
   |]
 
+-- Core.Queryable m BlockInfo (Core.EventsMatchingQuery BlockInfo) Core.SQLiteIndexer
```
Comment (berewt) [inline marconi-chain-index/src/Marconi/ChainIndex/Indexers/UtxoQuery.hs:293]: Is there a reason for this comment to be there?

File: marconi-core/src/Marconi/Core/Indexer/SQLiteAggregateQuery.hs [hunk 5]
```diff
@@ -98,6 +98,24 @@ mkSQLiteAggregateQuery sources = do
...
+          lastSyncPoint ix'
+     in lift . fmap minimum . traverse getPoint . Lens.view databases $ query
+
+  lastStablePoint query@(SQLiteAggregateQuery _ _) =
+    let getPoint :: SQLiteSourceProvider m point -> m point
+        getPoint (SQLiteSourceProvider ix) = do
+          ix' <- liftIO $ Con.readMVar ix
+          lastStablePoint ix'
+     in lift . fmap minimum . traverse getPoint . Lens.view databases $ query
+
```
Comment (berewt) [inline marconi-core/src/Marconi/Core/Indexer/SQLiteAggregateQuery.hs:118]: It looks good I think, no undedicability, no ambiguous types, great.

**Review by berewt — APPROVED** [review]
LGTM. Minor remarks. Regarding the formatting, let's continue with what the auto-formatting is doing.

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/UtxoQuery.hs (hunk 4 above)
Comment (willjgould) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers/UtxoQuery.hs:293]: That will have been a reminder to myself that I forgot to remove!

**Pull request merged**